File: templates/problems/05_binary_search/01_binary_search/typescript/solution.ts
export function solveSearch(nums: number[], target: number): number {
  // Binary search on a sorted array.
  // Time: O(log n), Space: O(1).
  let lo = 0;
  let hi = nums.length - 1;
  while (lo <= hi) {
    const mid = lo + Math.floor((hi - lo) / 2);
    if (nums[mid] === target) {
      return mid;
    } else if (nums[mid] < target) {
      lo = mid + 1;
    } else {
      hi = mid - 1;
    }
  }
  return -1;
}

File: templates/problems/05_binary_search/01_binary_search/python/solution.py
def solve_search(nums: list[int], target: int) -> int:
    """Binary search on a sorted array.

    Time: O(log n), Space: O(1).
    """
    lo, hi = 0, len(nums) - 1
    while lo <= hi:
        mid = lo + (hi - lo) // 2
        if nums[mid] == target:
            return mid
        elif nums[mid] < target:
            lo = mid + 1
        else:
            hi = mid - 1
    return -1
